firehose/firehose_kinesis.go

```diff
@@ -36,20 +40,31 @@ type FirehoseWriterConfig struct {
 	FlushSize int
 }
 
-func NewFirehoseWriter(config FirehoseWriterConfig, mockEndpoint string) (*FirehoseWriter, error) {
+// NewFirehoseWriter constructs a FirehoseWriter
+func NewFirehoseWriter(config FirehoseWriterConfig) (*FirehoseWriter, error) {
 	if config.FlushCount > 500 || config.FlushCount < 1 {
 		return nil, fmt.Errorf("FlushCount must be between 1 and 500 messages")
 	}
 	if config.FlushSize < 1 || config.FlushSize > 4*1024*1024 {
 		return nil, fmt.Errorf("FlushSize must be between 1 and 4*1024*1024 (4 Mb)")
 	}
-	return &FirehoseWriter{
-		conf:         &config,
-		mockEndpoint: mockEndpoint,
-		batchers:     map[string]batcher.Batcher{},
-	}, nil
+
+	sess := session.Must(session.NewSession(aws.NewConfig().WithRegion(config.Region).WithMaxRetries(4)))
+
+	f := &FirehoseWriter{
+		streamName:     config.StreamName,
+		firehoseClient: awsFirehose.New(sess),
```

Comment:
I'm somewhat partial to objects that use AWS taking in the service interface they use as an argument, e.g. in this case taking in [`firehoseiface.FirehoseAPI`](https://godoc.org/github.com/aws/aws-sdk-go/service/firehose/firehoseiface#FirehoseAPI). It makes testing a little bit easier, simplifies the config required to create the object, and pushes some environment-specific stuff like establishing a session (which has its own config requirements like credentials, etc.) out to users of the struct like `main()`, where I think it belongs.

(Also since AWS is adding context support to the SDK, there's a future world where we can wrap these interface types with tracing instrumentation, etc.)